### PR TITLE
fix(cli): keep default cc flags in build 

### PR DESCRIPTION
**Pull request by ObserverOfTime**
<details>

<summary>Complete commands from CI</summary>

#### release

```bash
i686-linux-gnu-gcc -O2 -ffunction-sections -fdata-sections -fPIC -m32 -march=i686 -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### debug

```bash
i686-linux-gnu-gcc -O0 -ffunction-sections -fdata-sections -fPIC -gdwarf-4 -fno-omit-frame-pointer -m32 -march=i686 -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -Wextra -Werror=implicit-function-declaration -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### release

```bash
powerpc64-linux-gnu-gcc -O2 -ffunction-sections -fdata-sections -fPIC -m64 -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### debug

```bash
powerpc64-linux-gnu-gcc -O0 -ffunction-sections -fdata-sections -fPIC -gdwarf-4 -fno-omit-frame-pointer -m64 -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -Wextra -Werror=implicit-function-declaration -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### release

```bash
arm-linux-gnueabi-gcc -O2 -ffunction-sections -fdata-sections -fPIC -march=armv6 -marm -mfloat-abi=soft -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### debug

```bash
arm-linux-gnueabi-gcc -O0 -ffunction-sections -fdata-sections -fPIC -gdwarf-4 -fno-omit-frame-pointer -march=armv6 -marm -mfloat-abi=soft -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -Wextra -Werror=implicit-function-declaration -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### release

```bash
aarch64-linux-gnu-gcc -O2 -ffunction-sections -fdata-sections -fPIC -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### debug

```bash
aarch64-linux-gnu-gcc -O0 -ffunction-sections -fdata-sections -fPIC -gdwarf-4 -fno-omit-frame-pointer -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -Wextra -Werror=implicit-function-declaration -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### release

```bash
cc -O2 -ffunction-sections -fdata-sections -fPIC -m64 -std=c11 -I /home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -shared /home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/parser.c /home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c -o /home/runner/work/tree-sitter/tree-sitter/parser.so
```

#### debug

```bash
cc -O0 -ffunction-sections -fdata-sections -fPIC -gdwarf-4 -fno-omit-frame-pointer -m64 -std=c11 -I /home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src -Wall -Wextra -Werror=implicit-function-declaration -shared /home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/parser.c /home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c -o /home/runner/work/tree-sitter/tree-sitter/parser.so
```

#### release

```
cc -O2 -ffunction-sections -fdata-sections -fPIC --target=arm64-apple-darwin -mmacosx-version-min=10.13 -std=c11 -I /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -dynamiclib -UTREE_SITTER_REUSE_ALLOCATOR /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/parser.c /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c -o /Users/runner/work/tree-sitter/tree-sitter/parser.dylib
```

#### debug

```bash
cc -O0 -ffunction-sections -fdata-sections -fPIC -gdwarf-2 -fno-omit-frame-pointer --target=arm64-apple-darwin -mmacosx-version-min=10.13 -std=c11 -I /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src -Wall -Wextra -Werror=implicit-function-declaration -dynamiclib -UTREE_SITTER_REUSE_ALLOCATOR /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/parser.c /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c -o /Users/runner/work/tree-sitter/tree-sitter/parser.dylib
```

#### release

```
cc -O2 -ffunction-sections -fdata-sections -fPIC -m64 --target=x86_64-apple-darwin -mmacosx-version-min=10.13 -std=c11 -I /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -dynamiclib -UTREE_SITTER_REUSE_ALLOCATOR /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/parser.c /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c -o /Users/runner/work/tree-sitter/tree-sitter/parser.dylib
```

#### debug

```bash
cc -O0 -ffunction-sections -fdata-sections -fPIC -gdwarf-2 -fno-omit-frame-pointer -m64 --target=x86_64-apple-darwin -mmacosx-version-min=10.13 -std=c11 -I /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src -Wall -Wextra -Werror=implicit-function-declaration -dynamiclib -UTREE_SITTER_REUSE_ALLOCATOR /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/parser.c /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c -o /Users/runner/work/tree-sitter/tree-sitter/parser.dylib
```

#### release

```cmd
C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x86\\cl.exe -nologo -MD -O2 -Brepro -std:c11 -I D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src -W4 -DNDEBUG -LD -utf-8 D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\parser.c D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\scanner.c -link -out:D:\\a\\tree-sitter\\tree-sitter\\parser.dll
```

#### debug

```cmd
C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x86\\cl.exe -nologo -MD -Z7 -Brepro -std:c11 -I D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src -W4 -LDd -utf-8 D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\parser.c D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\scanner.c -link -out:D:\\a\\tree-sitter\\tree-sitter\\parser.dll
```

#### release

```cmd
C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x64\\cl.exe -nologo -MD -O2 -Brepro -std:c11 -I D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src -W4 -DNDEBUG -LD -utf-8 D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\parser.c D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\scanner.c -link -out:D:\\a\\tree-sitter\\tree-sitter\\parser.dll
```

#### debug

```cmd
C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x64\\cl.exe -nologo -MD -Z7 -Brepro -std:c11 -I D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src -W4 -LDd -utf-8 D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\parser.c D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\scanner.c -link -out:D:\\a\\tree-sitter\\tree-sitter\\parser.dll
```

### aarch64-pc-windows-msvc

_Cannot run in CI_

</details>

_The `release` / `debug` variations need #3279_

### CODE REVIEW

File: cli/loader/src/lib.rs [hunk 1]
```diff
@@ -553,16 +554,38 @@ impl Loader {
...
-            .flag_if_supported("-Werror=implicit-function-declaration");
-        let compiler = cc.get_compiler();
+            .host(BUILD_HOST)
+            .file(&config.parser_path)
+            .includes(&config.header_paths);
+
+        if let Some(scanner_path) = config.scanner_path.as_ref() {
+            if scanner_path.extension() != Some("c".as_ref()) {
+                cc_config.cpp(true);
+                eprintln!("Warning: Using a C++ scanner is now deprecated. Please migrate your scanner code to C, as C++ support will be removed in the near future.");
```
Comment (ObserverOfTime) [inline cli/loader/src/lib.rs:569]: Is it time to drop this altogether yet?

**Comment by clason** [comment]
Further requests one might fold into this:
* create directories if not existing (i.e., `ts b -o foo/bar.so`)
* call `ts g` if `parser.c` doesn't exist

File: cli/loader/src/lib.rs (hunk 1 above)
Comment (clason) [reply cli/loader/src/lib.rs:569]: Nope.

**Comment by clason** [comment]
Out of interest (and for the sake of documentation): what flags are used by default (esp. `-O`)?

**Comment by ObserverOfTime** [comment]
>  * create directories if not existing (i.e., `ts b -o foo/bar.so`)
>  * call `ts g` if `parser.c` doesn't exist

Those are beyond my capabilities.

> * output parity for `make` and  `ts b`

We don't necessarily want that.
`ts b` is an easy and cross-platform way to build the parser.
`make` is mostly meant to be used in distro packages, which have their own preferred flags.

**Comment by ObserverOfTime** [comment]
Windows and macOS builds fail when `TREE_SITTER_REUSE_ALLOCATOR` is set. Maybe we need `TS_PUBLIC` again? :thinking:

**Comment by clason** [comment]
I think we do. If distro packages need their own flags, they will write their own build scripts

**Comment by ObserverOfTime** [comment]
> I think we do. If distro packages need their own flags, they will write their own build scripts

The required flags will still be the same.

**Comment by ObserverOfTime** [comment]
> Out of interest (and for the sake of documentation): what flags are used by default (esp. `-O`)?

The release build uses `-O2` and the debug build uses `-O0`. `CFLAGS` are applied after `-std`.

---

Complete commands from CI (release build):

#### i686-unknown-linux-gnu

```bash
i686-linux-gnu-gcc -O2 -ffunction-sections -fdata-sections -fPIC -m32 -march=i686 -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -g -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### powerpc64-unknown-linux-gnu

```bash
powerpc64-linux-gnu-gcc -O2 -ffunction-sections -fdata-sections -fPIC -m64 -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -g -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### arm-unknown-linux-gnueabi

```bash
arm-linux-gnueabi-gcc -O2 -ffunction-sections -fdata-sections -fPIC -march=armv6 -marm -mfloat-abi=soft -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -g -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### aarch64-unknown-linux-gnu

```bash
aarch64-linux-gnu-gcc -O2 -ffunction-sections -fdata-sections -fPIC -std=c11 -I /project/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -g -shared /project/test/fixtures/grammars/bash/src/parser.c /project/test/fixtures/grammars/bash/src/scanner.c -o /project/parser.so
```

#### x86_64-unknown-linux-gnu

```bash
cc -O2 -ffunction-sections -fdata-sections -fPIC -m64 -std=c11 -I /home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -g -shared /home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/parser.c /home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c -o /home/runner/work/tree-sitter/tree-sitter/parser.so
```

#### aarch64-apple-darwin

```
cc -O2 -ffunction-sections -fdata-sections -fPIC --target=arm64-apple-darwin -mmacosx-version-min=10.13 -std=c11 -I /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -g -dynamiclib -UTREE_SITTER_REUSE_ALLOCATOR /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/parser.c /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c -o /Users/runner/work/tree-sitter/tree-sitter/parser.dylib
```

#### x86_64-apple-darwin

```
cc -O2 -ffunction-sections -fdata-sections -fPIC -m64 --target=x86_64-apple-darwin -mmacosx-version-min=10.13 -std=c11 -I /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src -Wall -DNDEBUG -Werror=implicit-function-declaration -g -dynamiclib -UTREE_SITTER_REUSE_ALLOCATOR /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/parser.c /Users/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c -o /Users/runner/work/tree-sitter/tree-sitter/parser.dylib
```

#### i686-pc-windows-msvc

```cmd
C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x86\\cl.exe -nologo -MD -O2 -Brepro -std:c11 -I D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src -W4 -DNDEBUG /nologo /LD D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\parser.c D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\scanner.c /link /out:D:\\a\\tree-sitter\\tree-sitter\\parser.dll
```

#### x86_64-pc-windows-msvc

```cmd
C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x64\\cl.exe -nologo -MD -O2 -Brepro -std:c11 -I D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src -W4 -DNDEBUG /nologo /LD D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\parser.c D:\\a\\tree-sitter\\tree-sitter\\test/fixtures/grammars/bash\\src\\scanner.c /link /out:D:\\a\\tree-sitter\\tree-sitter\\parser.dll
```

#### aarch64-pc-windows-msvc

_Cannot run in CI_

**Comment by ObserverOfTime** [comment]
```
/home/runner/work/tree-sitter/tree-sitter/test/fixtures/grammars/bash/src/scanner.c:113:9: runtime error: null pointer passed as argument 2, which is declared to never be null
```

Adding `assert(heredoc->delimiter.contents != NULL);` before that line should silence UBSAN. @amaanq

**Comment by clason** [comment]
Would be good to document that somewhere, as a reference.

(nvim-treesitter/makefile uses `-Os`, but I don't see much difference in size)

**Comment by ObserverOfTime** [comment]
> Would be good to document that somewhere, as a reference.

I think the default flags are too complicated to document and not _that_ important.
We can always point to https://github.com/tree-sitter/tree-sitter/pull/3188#issuecomment-2020278490 if anyone asks.

**Pull request merged**